src/wp-admin/includes/plugin.php

```diff
@@ -1268,6 +1268,9 @@ function uninstall_plugin( $plugin ) {
 		 * @since 2.7.0
 		 */
 		do_action( "uninstall_{$file}" );
+
+		// Only remove it after the action finished to ensure the plugin stays uninstallable if the uninstall times out/crashes.
+		update_option( 'uninstall_plugins', $wp_internal_uninstallable_plugins_new_option_value );
```

Comment:
The same would also need to be added above, right after the `include_once WP_PLUGIN_DIR . '/' . dirname( $file ) . '/uninstall.php';` line.

Plugins can either have an uninstall callback or an `uninstall.php` file. This section here is only for the former.